File: app/src/main/java/net/wuqs/ontime/db/Alarm.java
package net.wuqs.ontime.db;

import android.annotation.SuppressLint;
import android.arch.persistence.room.ColumnInfo;
import android.arch.persistence.room.Entity;
import android.arch.persistence.room.Ignore;
import android.arch.persistence.room.PrimaryKey;
import android.content.Context;
import android.net.Uri;
import android.os.Parcel;
import android.os.Parcelable;

import java.lang.reflect.Field;
import java.util.Calendar;

@Entity(tableName = "alarms")
public class Alarm implements Parcelable {

    @PrimaryKey(autoGenerate = true)
    private long id;

    @ColumnInfo(name = "hour")
    private int hour;

    @ColumnInfo(name = "minute")
    private int minute;

    @ColumnInfo(name = "title")
    private String title;

    @ColumnInfo(name = "ringtone_uri")
    private Uri ringtoneUri;

    @ColumnInfo(name = "enabled")
    private boolean enabled = true;

    @ColumnInfo(name = "repeat_type")
    private int repeatType = 0;

    @ColumnInfo(name = "repeat_cycle")
    private int repeatCycle = 0;

    @ColumnInfo(name = "repeat_index")
    private int repeatIndex = 0;

    @ColumnInfo(name = "activate_date")
    private Calendar activateDate;

    @ColumnInfo(name = "next_occurrence")
    private Calendar nextOccurrence;

    /**
     * Checks if another {@link Alarm} has the same display summary as this {@link Alarm}.
     *
     * @param alarm the {@link Alarm} to check.
     * @return whether the two {@link Alarm}s have the same display summary.
     */
    public boolean sameDisplaySummaryAs(Alarm alarm) {
        return this.id == alarm.id
                && this.hour == alarm.hour
                && this.minute == alarm.minute
                && this.title.equals(alarm.title)
                && this.enabled == alarm.enabled
                && this.repeatType == alarm.repeatType
                && this.repeatCycle == alarm.repeatCycle
                && this.repeatIndex == alarm.repeatIndex
                && this.activateDate.getTimeInMillis() == alarm.activateDate.getTimeInMillis()
                && this.nextOccurrence.getTimeInMillis() == alarm.nextOccurrence.getTimeInMillis();
    }

    /**
     * Adds the {@link Alarm} to database.
     *
     * @param context the {@link Context} used to get the {@link AppDatabase} instance.
     * @return the {@link Alarm} added to database, with its id updated.
     */
    public Alarm addToDatabase(Context context) {
        this.id = AppDatabase.getInstance(context.getApplicationContext()).alarmDAO()
                .insert(this);
        return this;
    }

    public Alarm() {
        this.activateDate = Calendar.getInstance();
    }

    @Ignore
    public Alarm(int hour, int minute, Uri ringtoneUri) {
        this.hour = hour;
        this.minute = minute;
        this.ringtoneUri = ringtoneUri;
        this.activateDate = Calendar.getInstance();
    }

    @Ignore
    public Alarm(Parcel in) {
        this.id = in.readLong();
        this.hour = in.readInt();
        this.minute = in.readInt();
        this.title = in.readString();
        this.ringtoneUri = DataTypeConverter.toUri(in.readString());
        this.enabled = DataTypeConverter.toBoolean(in.readByte());
        this.repeatType = in.readInt();
        this.repeatCycle = in.readInt();
        this.repeatIndex = in.readInt();
        this.activateDate = DataTypeConverter.toCalendar(in.readLong());
        this.nextOccurrence = DataTypeConverter.toCalendar(in.readLong());
    }

    public Alarm(Alarm another) {
        copyFrom(another);
    }

    public void copyFrom(Alarm another) {
        this.id = another.id;
        this.hour = another.hour;
        this.minute = another.minute;
        this.title = another.title;
        this.ringtoneUri = Uri.parse(another.ringtoneUri.toString());
        this.enabled = another.enabled;
        this.repeatType = another.repeatType;
        this.repeatCycle = another.repeatCycle;
        this.repeatIndex = another.repeatIndex;
        this.activateDate = (Calendar) another.activateDate.clone();
        this.nextOccurrence = (Calendar) another.nextOccurrence.clone();
    }

    public long getId() {
        return id;
    }

    public void setId(long id) {
        this.id = id;
    }

    public int getHour() {
        return hour;
    }

    public void setHour(int hour) {
        this.hour = hour;
    }

    public int getMinute() {
        return minute;
    }

    public void setMinute(int minute) {
        this.minute = minute;
    }

    public String getTitle() {
        return title;
    }

    public void setTitle(String title) {
        this.title = title;
    }

    public Uri getRingtoneUri() {
        return ringtoneUri;
    }

    public void setRingtoneUri(Uri ringtoneUri) {
        this.ringtoneUri = ringtoneUri;
    }

    public boolean isEnabled() {
        return enabled;
    }

    public void setEnabled(boolean enabled) {
        this.enabled = enabled;
    }

    public int getRepeatType() {
        return repeatType;
    }

    public void setRepeatType(int repeatType) {
        this.repeatType = repeatType;
    }

    public int getRepeatCycle() {
        return repeatCycle;
    }

    public void setRepeatCycle(int repeatCycle) {
        this.repeatCycle = repeatCycle;
    }

    public int getRepeatIndex() {
        return repeatIndex;
    }

    public void setRepeatIndex(int repeatIndex) {
        this.repeatIndex = repeatIndex;
    }

    public Calendar getActivateDate() {
        return activateDate;
    }

    public void setActivateDate(Calendar activateDate) {
        this.activateDate = activateDate;
    }

    public Calendar getNextOccurrence() {
        return nextOccurrence;
    }

    public void setNextOccurrence(Calendar nextOccurrence) {
        this.nextOccurrence = nextOccurrence;
    }

    public static final Parcelable.Creator<Alarm> CREATOR = new Parcelable.Creator<Alarm>() {
        @Override
        public Alarm createFromParcel(Parcel source) {
            return new Alarm(source);
        }

        @Override
        public Alarm[] newArray(int size) {
            return new Alarm[size];
        }
    };

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeLong(id);
        dest.writeInt(hour);
        dest.writeInt(minute);
        dest.writeString(title);
        dest.writeString(DataTypeConverter.toString(ringtoneUri));
        dest.writeByte(DataTypeConverter.toByte(enabled));
        dest.writeInt(repeatType);
        dest.writeInt(repeatCycle);
        dest.writeInt(repeatIndex);
        dest.writeLong(DataTypeConverter.toLong(activateDate));
        dest.writeLong(DataTypeConverter.toLong(nextOccurrence));
    }

    @SuppressLint("DefaultLocale")
    @Override
    public String toString() {
        return String.format("{id=%d, %02d:%02d, title=%s, enabled=%b, " +
                        "repeatType=%d, repeatCycle=%d, repeatIndex=%d, " +
                        "activateDate=%s, next=%s}",
                id, hour, minute, title, enabled,
                repeatType, repeatCycle, repeatIndex,
                activateDate.getTime().toString(), nextOccurrence.getTime().toString());
    }
}

File: app/src/main/java/net/wuqs/ontime/db/Alarm.kt
package net.wuqs.ontime.db

import android.content.Context
import android.content.Intent
import android.net.Uri
import android.os.Bundle
import android.os.Parcel
import android.os.Parcelable
import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.Ignore
import androidx.room.PrimaryKey
import net.wuqs.ontime.alarm.*
import net.wuqs.ontime.feature.editalarm.binString
import net.wuqs.ontime.feature.editalarm.hexString
import net.wuqs.ontime.util.readBoolean
import net.wuqs.ontime.util.writeBoolean
import java.util.*

/**
 * @property activateDate the date when this Alarm is activated; for a past alarm, this represents
 *     when this Alarm went off
 * @property nextTime the exact time when this Alarm will go off next time; for a past alarm, this
 *     represents when this Alarm was dismissed/completed
 */
@Entity(tableName = "alarms")
class Alarm(
    @PrimaryKey var id: Long = INVALID_ID,
    @ColumnInfo(name = Columns.HOUR) var hour: Int = 0,
    @ColumnInfo(name = Columns.MINUTE) var minute: Int = 0,
    @ColumnInfo(name = Columns.TITLE) var title: String? = "",
    @ColumnInfo(name = Columns.RINGTONE_URI) var ringtoneUri: Uri? = null,
    @ColumnInfo(name = Columns.VIBRATE) var vibrate: Boolean = false,
    @ColumnInfo(name = Columns.SILENCE_AFTER) var silenceAfter: Int = -1,
    @ColumnInfo(name = Columns.ENABLED) var isEnabled: Boolean = true,
    @ColumnInfo(name = Columns.REPEAT_TYPE) var repeatType: Int = 0,
    @ColumnInfo(name = Columns.REPEAT_CYCLE) var repeatCycle: Int = 0,
    @ColumnInfo(name = Columns.REPEAT_INDEX) var repeatIndex: Int = 0,
    @ColumnInfo(name = Columns.ACTIVATE_DATE) var activateDate: Calendar? = Calendar.getInstance(),
    @ColumnInfo(name = Columns.NEXT_OCCURRENCE) var nextTime: Calendar? = null,
    @ColumnInfo(name = Columns.SNOOZED) var snoozed: Int = 0,
    @ColumnInfo(name = Columns.NOTES) var notes: String = "",
    @ColumnInfo(name = Columns.HISTORICAL) var isHistorical: Boolean = false,
    @ColumnInfo(name = Columns.PARENT_ALARM_ID) var parentAlarmId: Long? = null
) : Parcelable {

    object Columns {
        const val ID = "id"
        const val HOUR = "hour"
        const val MINUTE = "minute"
        const val TITLE = "title"
        const val RINGTONE_URI = "ringtone_uri"
        const val VIBRATE = "vibrate"
        const val SILENCE_AFTER = "silence_after"
        const val ENABLED = "enabled"
        const val REPEAT_TYPE = "repeat_type"
        const val REPEAT_CYCLE = "repeat_cycle"
        const val REPEAT_INDEX = "repeat_index"
        const val ACTIVATE_DATE = "activate_date"
        const val NEXT_OCCURRENCE = "next_occurrence"
        const val SNOOZED = "snoozed"
        const val NOTES = "notes"
        const val HISTORICAL = "historical"
        const val PARENT_ALARM_ID = "parent_alarm_id"
    }

    companion object {

        const val ALARM_ID = "net.wuqs.ontime.extra.ALARM_ID"

        /*
        Repeat types
        1st digit: display type
        2nd digit: calculate type
        3rd digit: flag for repeat yearly
         */
        const val NON_REPEAT = 0x0
        const val REPEAT_DAILY = 0x11
        const val REPEAT_WEEKLY = 0x22
        const val REPEAT_MONTHLY_BY_DATE = 0x43
        const val REPEAT_MONTHLY_BY_WEEK = 0x83
        const val REPEAT_YEARLY_BY_DATE = 0x144
        const val REPEAT_YEARLY_BY_WEEK = 0x184

        const val INVALID_ID = 0L

        @Suppress("unused")
        @JvmField
        val CREATOR: Parcelable.Creator<Alarm> = object : Parcelable.Creator<Alarm> {
            override fun createFromParcel(source: Parcel): Alarm = Alarm(source)
            override fun newArray(size: Int): Array<Alarm?> = arrayOfNulls(size)
        }
    }

    init {
        if (!isHistorical) activateDate?.setHms(0)
    }

    // TODO: Move to AlarmUtil.kt
    /**
     * Determines the [Alarm]'s next occurrence.
     */
    fun getNextOccurrence(now: Calendar = Calendar.getInstance()): Calendar? = when (repeatType) {
        NON_REPEAT -> nextTimeNonRepeat(now)
        REPEAT_DAILY -> nextTimeDaily(now)
        REPEAT_WEEKLY -> nextTimeWeekly(now)
        REPEAT_MONTHLY_BY_DATE -> nextTimeMonthlyByDate(now)
        REPEAT_YEARLY_BY_DATE -> nextTimeYearlyByDate(now)
        else -> null
    }

    // TODO: Move to AlarmUtil.kt
    fun createAlarmStartIntent(context: Context): Intent {
        val bundle = Bundle().apply { putParcelable(EXTRA_ALARM_INSTANCE, this@Alarm) }
        return Intent(context, AlarmService::class.java).apply {
            action = ACTION_ALARM_START
            addCategory("ALARM_MANAGER")
            putExtra(EXTRA_ALARM_INSTANCE, bundle)
        }
    }

    @Ignore
    private constructor(source: Parcel) : this(
            id = source.readLong(),
            hour = source.readInt(),
            minute = source.readInt(),
            title = source.readString(),
            ringtoneUri = source.readParcelable(Uri::class.java.classLoader),
            vibrate = source.readInt().toBoolean(),
            silenceAfter = source.readInt(),
            isEnabled = source.readInt().toBoolean(),
            repeatType = source.readInt(),
            repeatCycle = source.readInt(),
            repeatIndex = source.readInt(),
            activateDate = source.readLong().toCalendar(),
            nextTime = source.readLong().takeIf { it != -1L }.toCalendar(),
            snoozed = source.readInt(),
            notes = source.readString()!!,
            isHistorical = source.readBoolean(),
            parentAlarmId = source.readLong().takeIf { it != INVALID_ID }
    )

    @Ignore
    constructor(another: Alarm) : this(
            id = another.id,
            hour = another.hour,
            minute = another.minute,
            title = another.title,
            ringtoneUri = another.ringtoneUri,
            vibrate = another.vibrate,
            silenceAfter = another.silenceAfter,
            isEnabled = another.isEnabled,
            repeatType = another.repeatType,
            repeatCycle = another.repeatCycle,
            repeatIndex = another.repeatIndex,
            activateDate = another.activateDate!!.clone() as Calendar,
            nextTime = another.nextTime?.clone() as Calendar?,
            snoozed = another.snoozed,
            notes = another.notes,
            isHistorical = another.isHistorical,
            parentAlarmId = another.parentAlarmId
    )

    override fun describeContents() = 0

    override fun writeToParcel(dest: Parcel, flags: Int) = dest.run {
        writeLong(id)
        writeInt(hour)
        writeInt(minute)
        writeString(title)
        writeParcelable(ringtoneUri, 0)
        writeInt(vibrate.toInt())
        writeInt(silenceAfter)
        writeInt(isEnabled.toInt())
        writeInt(repeatType)
        writeInt(repeatCycle)
        writeInt(repeatIndex)
        writeLong(activateDate.toLong()!!)
        writeLong(nextTime.toLong() ?: -1L)
        writeInt(snoozed)
        writeString(notes)
        writeBoolean(isHistorical)
        writeLong(parentAlarmId ?: INVALID_ID)
    }

    override fun toString(): String = if (!isHistorical) {
        "{id=$id, $hour:$minute, " +
                "title=$title, isEnabled=$isEnabled, " +
                "ringtone=$ringtoneUri, vibrate=$vibrate, silenceAfter=$silenceAfter, " +
                "repeatType=${repeatType.hexString}, repeatCycle=$repeatCycle, " +
                "repeatIndex=${repeatIndex.binString}, " +
                "activate=${activateDate?.time}, next=${nextTime?.time}, " +
                "snoozed=$snoozed, notes=$notes}"
    } else {
        "{historical, id=$id, parent=$parentAlarmId, $hour:$minute, title=$title, " +
                "activate=${activateDate?.time}, next=${nextTime?.time}, notes=$notes}"
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Alarm

        if (id != other.id) return false
        if (hour != other.hour) return false
        if (minute != other.minute) return false
        if (title != other.title) return false
        if (ringtoneUri != other.ringtoneUri) return false
        if (vibrate != other.vibrate) return false
        if (silenceAfter != other.silenceAfter) return false
        if (isEnabled != other.isEnabled) return false
        if (repeatType != other.repeatType) return false
        if (repeatCycle != other.repeatCycle) return false
        if (repeatIndex != other.repeatIndex) return false
        if (activateDate != other.activateDate) return false
        if (nextTime != other.nextTime) return false
        if (snoozed != other.snoozed) return false
        if (notes != other.notes) return false
        if (isHistorical != other.isHistorical) return false
        if (parentAlarmId != other.parentAlarmId) return false

        return true
    }

    override fun hashCode() = id.hashCode()
}
